Duplicated pki mount entries, vault unable to start

**Describe the bug**
This is the concise bug description after a lot of investigation - see additional context for detailed story.
Special cases can cause a duplication of PKI mounts at same path. The PKI secret engine is used by clients with active certs and leases. The secret engine will be deleted with terraform destroy. The destroy takes up to 1 min for cleanup tasks and succeeds. Afterwards, the PKI secret engine is recreated with terraform apply. A vault server restart will now fail with error for duplicated mount:
```
2019-05-16T14:37:07.721Z [ERROR] core: failed to mount entry: path=pki-tsc-max-kube/ error="cannot mount under existing mount "pki-tsc-max-kube/""
```
This is a dangerous bug, because the problem is not discovered during runtime. Only a vault restart discovers the issue. The delayed occurrence of the error can cause massive data loss, because a backup is required before the duplicated mount has been created.

**To Reproduce**
Steps to reproduce the behavior:
1. Create a PKI secret engine and role definition with terraform.
2. Run clients creating several active certs/keys with leases (~1,000)
3. Delete the PKI secret engine with terraform destroy. The destroy operation will take up to 1 minute cleaning up all leases etc. Terraform destroy will succeed, but vault server shows error message like:
```
05.04.19, 10:24:15.373  10.52.187.109   vault           ERROR   secrets.system.system_ca91d9d4: unmount failed: path=pki-tsc-max-kube/ error="Delete https://127.0.0.1:8500/v1/kv/vault/logical/3c99d546-1b6a-437e-b5fe-b1e77f02bba7/revoked/58-e4-fb-ca-a8-98-59-8a-9d-9a-18-9c-25-27-9f-0c-15-45-45-12: context canceled"
05.04.19, 10:24:15.372  10.52.187.109   vault           ERROR   core: failed to clear view for path being unmounted: error="Delete https://127.0.0.1:8500/v1/kv/vault/logical/3c99d546-1b6a-437e-b5fe-b1e77f02bba7/revoked/58-e4-fb-ca-a8-98-59-8a-9d-9a-18-9c-25-27-9f-0c-15-45-45-12: context canceled" path=pki-tsc-max-kube/
```
4. Recreate the PKI secret engine with terraform apply. The create will succeed.
5. Restart the vault server and startup will fail with error:
```
2019-05-16T14:37:07.721Z [ERROR] core: failed to mount entry: path=pki-tsc-max-kube/ error="cannot mount under existing mount "pki-tsc-max-kube/""
```

**Expected behavior**
1. Vault must always fail creation of duplicated secret engine mounts at same path. Even if the secret engine delete operation failed and is leaving inconsistent mount points.
2. Terraform PKI  secret engine destroy must fail if vault delete operation is not successful (this could also be a bug in the terraform vault provider).

**Environment:**
* Vault Server Version: 1.1.2 (docker container)
* Consul Server Version: 1.4.2 (docker container)
* Terraform v0.11.14 + provider.vault v1.8.0

Vault server configuration file(s):

```hcl
backend "consul" {
  address = "######"
  scheme = "https"
  path = "vault/"
  token = "###########"
  tls_cert_file = "/#####/####.pem"
  tls_key_file = "/#####/####.pem"
  tls_ca_file = "/#####/####.pem"
}

ui = "true"

seal "awskms" {
  region = "#############"
  kms_key_id = "###########"
}

listener "tcp" {
  address = "######"
  cluster_address = "######"
  tls_cert_file = "/#####/####.pem"
  tls_key_file = "/#####/####.pem"
}

telemetry {
  prometheus_retention_time = "24h",
  disable_hostname = true
}

api_addr = "https://#####"
```

**Additional context**
The error occurred during a vault upgrade from version 1.0.2 to 1.1.2. The vault server start failed with error:
```
2019-05-16T14:37:07.721Z [ERROR] core: failed to mount entry: path=pki-tsc-max-kube/ error="cannot mount under existing mount "pki-tsc-max-kube/""
```
First attempt was to restore a backup before the update started and using old vault server version. But the error still occurred. In the next steps older backups have been tried until **a 1 month old one** worked. Consequently, the issue must not be related to the vault upgrade.
Based on the error we assumed a PKI mount delete and create operation must have caused the problem. We analyzed vault audit log and found related mount operations. Based on this, we tested the expected last working and first failing backup, that proved our thesis. 
Now I reproduced the issue with assumption, heavy cleanup of certs and leases must cause the problem during delete.

--- a/vault/auth.go
+++ b/vault/auth.go
@@ -76,7 +76,7 @@ func (c *Core) enableCredentialInternal(ctx context.Context, entry *MountEntry, 
 	// Populate cache
 	NamespaceByID(ctx, ns.ID, c)
 
-	// Look for matching name
+	// Basic check for matching names
 	for _, ent := range c.auth.Entries {
 		if ns.ID == ent.NamespaceID {
 			switch {
@@ -85,7 +85,7 @@ func (c *Core) enableCredentialInternal(ctx context.Context, entry *MountEntry, 
 			case strings.HasPrefix(ent.Path, entry.Path):
 				fallthrough
 			case strings.HasPrefix(entry.Path, ent.Path):
-				return logical.CodedError(409, "path is already in use")
+				return logical.CodedError(409, fmt.Sprintf("path is already in use at %s", ent.Path))
 			}
 		}
 	}
@@ -95,6 +95,7 @@ func (c *Core) enableCredentialInternal(ctx context.Context, entry *MountEntry, 
 		return fmt.Errorf("token credential backend cannot be instantiated")
 	}
 
+	// Check for conflicts according to the router
 	if conflict := c.router.MountConflict(ctx, credentialRoutePrefix+entry.Path); conflict != "" {
 		return logical.CodedError(409, fmt.Sprintf("existing mount at %s", conflict))
 	}
@@ -258,11 +259,6 @@ func (c *Core) disableCredentialInternal(ctx context.Context, path string, updat
 		backend.Cleanup(ctx)
 	}
 
-	// Unmount the backend
-	if err := c.router.Unmount(ctx, path); err != nil {
-		return err
-	}
-
 	viewPath := entry.ViewPath()
 	switch {
 	case !updateStorage:
@@ -281,6 +277,11 @@ func (c *Core) disableCredentialInternal(ctx context.Context, path string, updat
 
 	// Remove the mount table entry
 	if err := c.removeCredEntry(ctx, strings.TrimPrefix(path, credentialRoutePrefix), updateStorage); err != nil {
+		return err
+	}
+
+	// Unmount the backend
+	if err := c.router.Unmount(ctx, path); err != nil {
 		return err
 	}
 

--- a/vault/external_tests/router/router_ext_test.go
+++ b/vault/external_tests/router/router_ext_test.go
@@ -5,19 +5,23 @@ import (
 
 	"github.com/hashicorp/vault/api"
 	"github.com/hashicorp/vault/builtin/credential/userpass"
+	"github.com/hashicorp/vault/builtin/logical/pki"
 	vaulthttp "github.com/hashicorp/vault/http"
 	"github.com/hashicorp/vault/sdk/logical"
 	"github.com/hashicorp/vault/vault"
 )
 
 func TestRouter_MountSubpath_Checks(t *testing.T) {
-	testRouter_MountSubpath(t, []string{"auth/abcd/123", "abcd/123"})
-	testRouter_MountSubpath(t, []string{"abcd/123", "auth/abcd/123"})
-	testRouter_MountSubpath(t, []string{"auth/abcd/123", "abcd/123"})
+	testRouter_MountSubpath(t, []string{"a/abcd/123", "abcd/123"})
+	testRouter_MountSubpath(t, []string{"abcd/123", "a/abcd/123"})
+	testRouter_MountSubpath(t, []string{"a/abcd/123", "abcd/123"})
 }
 
 func testRouter_MountSubpath(t *testing.T, mountPoints []string) {
 	coreConfig := &vault.CoreConfig{
+		LogicalBackends: map[string]logical.Factory{
+			"pki": pki.Factory,
+		},
 		CredentialBackends: map[string]logical.Factory{
 			"userpass": userpass.Factory,
 		},
@@ -31,22 +35,33 @@ func testRouter_MountSubpath(t *testing.T, mountPoints []string) {
 	vault.TestWaitActive(t, cluster.Cores[0].Core)
 	client := cluster.Cores[0].Client
 
+	// Test auth
 	authInput := &api.EnableAuthOptions{
 		Type: "userpass",
 	}
-
 	for _, mp := range mountPoints {
-		t.Logf("mounting %s", mp)
+		t.Logf("mounting %s", "auth/"+mp)
 		var err error
-		err = client.Sys().EnableAuthWithOptions(mp, authInput)
+		err = client.Sys().EnableAuthWithOptions("auth/"+mp, authInput)
+		if err != nil {
+			t.Fatalf("err: %v", err)
+		}
+	}
+
+	// Test secrets
+	mountInput := &api.MountInput{
+		Type: "pki",
+	}
+	for _, mp := range mountPoints {
+		t.Logf("mounting %s", "s/"+mp)
+		var err error
+		err = client.Sys().Mount("s/"+mp, mountInput)
 		if err != nil {
 			t.Fatalf("err: %v", err)
 		}
 	}
 
 	cluster.EnsureCoresSealed(t)
-
 	cluster.UnsealCores(t)
-
 	t.Logf("Done: %#v", mountPoints)
 }

--- a/vault/mount.go
+++ b/vault/mount.go
@@ -395,7 +395,21 @@ func (c *Core) mountInternal(ctx context.Context, entry *MountEntry, updateStora
 	// Ensure the cache is populated, don't need the result
 	NamespaceByID(ctx, ns.ID, c)
 
-	// Verify there are no conflicting mounts
+	// Basic check for matching names
+	for _, ent := range c.mounts.Entries {
+		if ns.ID == ent.NamespaceID {
+			switch {
+			// Existing is oauth/github/ new is oauth/ or
+			// existing is oauth/ and new is oauth/github/
+			case strings.HasPrefix(ent.Path, entry.Path):
+				fallthrough
+			case strings.HasPrefix(entry.Path, ent.Path):
+				return logical.CodedError(409, fmt.Sprintf("path is already in use at %s", ent.Path))
+			}
+		}
+	}
+
+	// Verify there are no conflicting mounts in the router
 	if match := c.router.MountConflict(ctx, entry.Path); match != "" {
 		return logical.CodedError(409, fmt.Sprintf("existing mount at %s", match))
 	}
@@ -571,11 +585,6 @@ func (c *Core) unmountInternal(ctx context.Context, path string, updateStorage b
 		backend.Cleanup(ctx)
 	}
 
-	// Unmount the backend entirely
-	if err := c.router.Unmount(ctx, path); err != nil {
-		return err
-	}
-
 	viewPath := entry.ViewPath()
 	switch {
 	case !updateStorage:
@@ -592,9 +601,15 @@ func (c *Core) unmountInternal(ctx context.Context, path string, updateStorage b
 			return err
 		}
 	}
+
 	// Remove the mount table entry
 	if err := c.removeMountEntry(ctx, path, updateStorage); err != nil {
 		c.logger.Error("failed to remove mount entry for path being unmounted", "error", err, "path", path)
+		return err
+	}
+
+	// Unmount the backend entirely
+	if err := c.router.Unmount(ctx, path); err != nil {
 		return err
 	}
 

--- a/vault/mount_test.go
+++ b/vault/mount_test.go
@@ -248,6 +248,11 @@ func TestCore_Unmount(t *testing.T) {
 }
 
 func TestCore_Unmount_Cleanup(t *testing.T) {
+	testCore_Unmount_Cleanup(t, false)
+	testCore_Unmount_Cleanup(t, true)
+}
+
+func testCore_Unmount_Cleanup(t *testing.T, causeFailure bool) {
 	noop := &NoopBackend{}
 	c, _, root := TestCoreUnsealed(t)
 	c.logicalBackends["noop"] = func(context.Context, *logical.BackendConfig) (logical.Backend, error) {
@@ -304,8 +309,17 @@ func TestCore_Unmount_Cleanup(t *testing.T) {
 		t.Fatalf("bad: %#v", resp)
 	}
 
+	if causeFailure {
+		view.(*BarrierView).setReadOnlyErr(logical.ErrSetupReadOnly)
+	}
+
 	// Unmount, this should cleanup
-	if err := c.unmount(namespace.RootContext(nil), "test/"); err != nil {
+	err = c.unmount(namespace.RootContext(nil), "test/")
+	switch {
+	case err != nil && causeFailure:
+	case err == nil && causeFailure:
+		t.Fatal("expected error")
+	case err != nil:
 		t.Fatalf("err: %v", err)
 	}
 
@@ -327,8 +341,23 @@ func TestCore_Unmount_Cleanup(t *testing.T) {
 	if err != nil {
 		t.Fatalf("err: %v", err)
 	}
-	if len(out) != 0 {
+	switch {
+	case len(out) == 1 && causeFailure:
+	case len(out) == 0 && causeFailure:
+		t.Fatal("expected a value")
+	case len(out) != 0:
 		t.Fatalf("bad: %#v", out)
+	case !causeFailure:
+		return
+	}
+
+	// At this point just in the failure case, check mounting
+	if err := c.mount(namespace.RootContext(nil), me); err == nil {
+		t.Fatal("expected error")
+	} else {
+		if !strings.Contains(err.Error(), "path is already in use at") {
+			t.Fatalf("expected a path is already in use error, got %v", err)
+		}
 	}
 }
 

--- a/vault/router.go
+++ b/vault/router.go
@@ -121,7 +121,7 @@ func (r *Router) Mount(backend logical.Backend, prefix string, mountEntry *Mount
 
 	// Create a mount entry
 	re := &routeEntry{
-		tainted:       false,
+		tainted:       mountEntry.Tainted,
 		backend:       backend,
 		mountEntry:    mountEntry,
 		storagePrefix: storageView.Prefix(),

